### PR TITLE
Iroha2 python initial commit

### PR DESCRIPTION
**Pull request by i1i1**
Signed-off-by: i1i1 <vanyarybin1@live.ru>

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1,201 +0,0 @@
-                                 Apache License
```
Comment (eadventurous) [inline LICENSE:1]: I think we need license, can you consult with @LiraLemur ?

File: iroha_python/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "iroha_python"
+version = "0.1.0"
+authors = ["Iroha 2 team <https://github.com/orgs/soramitsu/teams/iroha2>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+[lib]
+name = "iroha2_sys"
```
Comment (eadventurous) [inline iroha_python/Cargo.toml:9]: Maybe `iroha2_bindings`?

File: iroha_python/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,195 @@
+// Allow panic because of bad and unsafe pyo3
```
Comment (eadventurous) [inline iroha_python/src/lib.rs:1]: I think you need to add some top level docs to this lib and `generator` lib and `iroha2_sys`. Right now it is a bit difficult to understand exact functions of each of them.

File: iroha_python/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    #[allow(clippy::missing_const_for_fn)]
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+impl<'source, T: serde::de::DeserializeOwned> FromPyObject<'source> for Dict<T> {
+    fn extract(obj: &'source PyAny) -> PyResult<Self> {
+        let obj = if obj.hasattr("to_rust")? {
+            obj.call_method0("to_rust")?
```
Comment (eadventurous) [inline iroha_python/src/lib.rs:94]: It would be good to have "to_rust" str as a constant, it is referenced in several places.

**Comment by eadventurous** [comment]
In general I would suggest to add following things to readme:

1. Architecture of the lib (custom generator, pyo3, pythonize relationship) and why these choices were made
2. How to update the library section, what can be changed, what should be generated

File: iroha_python/Cargo.toml (hunk 2 above)
Comment (i1i1) [reply iroha_python/Cargo.toml:9]: I kinda like naming with `_sys`, as it is very rusty

File: iroha_python/src/lib.rs (hunk 4 above)
Comment (i1i1) [reply iroha_python/src/lib.rs:94]: It is used in single place here. You won't be able to synchronize this contract with python anyway, so i think there is no need for constant.

Comment (i1i1) [reply iroha_python/src/lib.rs:94]: I mean it will anyway be soft contract and it is referenced only in 2 lines which are tighed up together.

Comment (eadventurous) [reply iroha_python/src/lib.rs:94]: Yes, I see, the other one was in python

Comment (eadventurous) [reply iroha_python/src/lib.rs:94]: Ok

File: README.md [hunk 5]
```diff
@@ -1,2 +1,52 @@
...
+cargo build --release
+cp target/release/libiroha2_sys.so iroha2/sys/iroha2.so
+python3 test.py
+```
+
+## Generating python sys sources
+
+```sh
+cd iroha_python/generate
+cargo run ../iroha2/sys
```
Comment (eadventurous) [inline README.md:20]: Shouldn't you add args after `--`?
`cargo run -- ../iroha2/sys` maybe?

File: README.md [hunk 6]
```diff
@@ -1,2 +1,52 @@
...
+
+```sh
+cd iroha_python
+cargo build --release
+cp target/release/libiroha2_sys.so iroha2/sys/iroha2.so
+python3 test.py
+```
+
+## Generating python sys sources
+
```
Comment (eadventurous) [inline README.md:17]: Can you mention that you need to regenerate code whenever you update _?

File: README.md (hunk 5 above)
Comment (i1i1) [reply README.md:20]: No it is not needed, as `cargo run` never takes arguments without flags

**Review by eadventurous — APPROVED** [review]

**Pull request merged**